### ISSUE TITLE
TWIG function checkAccess fails when not logged in

### ISSUE THREAD

**Issue by alani1** [userfrosting/UserFrosting#736]
I check for certain roles in FE using the Twig Function checkAccess (defined in sprinkles/account/src/Twig/AccountExtension.php (line 42). When User is not logged in, the result is below error message.

Slim Application Error:\nType: TypeError\nMessage: Argument 1 passed to UserFrosting\\Sprinkle\\Account\\Authorize\\AuthorizationManager::checkAccess() must be an instance of UserFrosting\\Sprinkle\\Account\\Model\\User, null given

can we add an additional check if currentUser is not null ? With this I could avoid an additional if block calling checkAuthenticated. E.g. in checkAccess simply use:

    if ( $currentUser ) {
       return $authorizer->checkAccess($currentUser, $slug, $params);
    } else {
       return false;
    }

**Comment by alexweissman** [comment]
I'm not sure at which level we should handle this.

I was looking at the [OWASP Top Ten Cheat Sheet](https://www.owasp.org/index.php/OWASP_Top_Ten_Cheat_Sheet) recently, and item A7, "Missing Function Level Access Control", suggests that:

> Ensure every page requires a role, even if it is "guest".

So I'm wondering if maybe we _should_ create a "guest" user object after all?  If we did that, your guest users could fail the access check by virtue of not having the "User" role.

**Comment by lcharette** [comment]
+1 for @alexweissman thing. Could be useful to handle what a guest can do and change it overtime, especially when dealing with role attached to multiple entities  :  Guess can comment on project A post, but not on project B (and can't even know project C exist). 

(Still don't solve original issue :P)

**Comment by alexweissman** [comment]
Well it would solve the original issue because then `$currentUser` would always be a `User` object.

**Comment by lcharette** [comment]
Right

**Comment by alexweissman** [comment]
Oh, also, @alani1 , are you adding `->add('authGuard');` to the end of your route definitions for these pages?

**Comment by alani1** [comment]
@alexweissman, the route is declared in core sprinkle and there it does not use ->add('authGuard'); When I overrite the route and add ('authGuard'), the anonymous or guest users get a AuthExpiredException which (if I remember right) should redirect to the login-page on production systems.

I agree probably the guest user would be a good solution.  In fact in the function comment for user() in Authenticator it already says: 
'* Try to get the currently authenticated user, returning a guest user if none was found.'
Just the implementation is missing

**Comment by alexweissman** [comment]
Ok, let's consider this in a revision once we release 4.1.

**Comment by alexweissman** [comment]
It would seem that there is another issue that having a "guest user" would solve: `AuthorizationManager`'s dependency on `Authenticator`.  Right now we have this in `AuthorizationManager`:

```
        if ($this->ci->authenticator->guest()) {
            if ($debug) {
                $this->ci->authLogger->debug("User is not logged in.  Access denied.");
            }
            return false;
        }
```

If we could simply test the `user` object directly to determine if it were a guest or not, this would reduce coupling and make `AuthorizationManager` not dependent on `Authenticator`.  This, in turn, would make it easier for people to implement their own custom versions of `Authenticator`.

**Comment by lcharette** [comment]
RE @alexweissman comment right above, testing is the user is a guest in `checkAccess` doesn't make sense at all. 

Since the user tested in `checkaccess` is the one passed in argument, it doesn't matter if it's a guest or not. Sure, 90% of the time, we test the `currentUser`, which can be a guest, but we also might want to test another user. By doing so, `User A` is testing to see if `User B` has permissions to do whatever. `User B` will always be a "guest" in the eyes of `User A`, and the guest status will always be checked against the `User A` in the above code. 

Long story short, as long as `checkaccess` check the user in the argument, we don't need to check if said user is a guest or not. This should be handled in `currentUser`... or `checkaccess` should always use `currentUser` and not accept any `user` arguments.

**Comment by lcharette** [comment]
As for the original issue, `checkAccess` now accept a `null` user, as of 5be1a3779979a2c9facfbb294bc41bf15d6d75de. Further discussion about guest user can be done in #914.
